File: javascript/medium/maximumSizeSubarraySumEqualsK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxSubArrayLen = function(nums, k) {
    let maxLen = 0, sum = 0, map = new Map();
    for(let i = 0; i < nums.length; i++) {
        sum += nums[i];
        if(sum === k) {
            maxLen = Math.max(maxLen, (i+1));
        } else if(map.has(sum - k)) {
            maxLen = Math.max(maxLen, (i - map.get(sum-k)));
        }
        if(!map.has(sum)) {
            map.set(sum, i);    
        }
    }
    return maxLen;
};

File: ruby/medium/maximum_size_subarray_sum_equals_k.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def max_sub_array_len(nums, k)
    map = {}
    sum, max_len = 0, 0
    nums.each_with_index do |num, i|
        sum += num
        if sum == k
            max_len = [max_len, i+1].max
        elsif map.has_key?(sum-k)
            max_len = [max_len, (i - map[sum-k])].max
        end
        map[sum] = i if !map.has_key?(sum)
    end
    return max_len
end
